### PR TITLE
derive new addresses from xpubs when balances are refreshed

### LINKED ISSUE #1830: Refreshing balances should derive new xpubkey addresses and check to see if there were deposits
## Abstract

When you refresh balances, you should check to see if there were coins sent to new addresses of xpubkeys

## Motivation

People with xpubkeys will occasionally deposit coins to new derived addresses. It's important that refreshing balances catches those.

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #1830

### CODE REVIEW

File: rotkehlchen/chain/manager.py [hunk 1]
```diff
@@ -621,10 +621,18 @@ def query_balances(
             self.query_polkadot_balances(ignore_cache=ignore_cache)
         if should_query_avax:
             self.query_avax_balances(ignore_cache=ignore_cache)
+        if ignore_cache is True:
+            self.derive_new_addresses_from_xpubs()
```
Comment (yabirgb) [inline rotkehlchen/chain/manager.py:625]: Shouldn't this happen before btc and bch balances are queried? Also this should happen only if BTC or BCH are queried

File: rotkehlchen/chain/manager.py [hunk 2]
```diff
@@ -621,10 +621,18 @@ def query_balances(
...
 
         self.totals = self.balances.recalculate_totals()
         return self.get_balances_update()
 
+    def derive_new_addresses_from_xpubs(self) -> None:
+        """Derive new addresses from xpubs."""
+        xpub_manager = XpubManager(chain_manager=self)
+        xpub_manager.check_for_new_xpub_addresses(blockchain=SupportedBlockchain.BITCOIN)
+        xpub_manager.check_for_new_xpub_addresses(blockchain=SupportedBlockchain.BITCOIN_CASH)
```
Comment (yabirgb) [inline rotkehlchen/chain/manager.py:634]: I believe it would be better to not combine this two. Correct me if I'm wrong but the derivation process makes some external calls and those can be slow. If you are querying only bch balances you would also be deriving BTC addresses and that might not be needed. Also think in accounts with a lot of addresses

File: docs/changelog.rst [hunk 3]
```diff
@@ -2,6 +2,7 @@
 Changelog
 =========
 
+* :feature:`1830` New addresses are now derived from XPUBs when balances are refreshed.
```
Comment (yabirgb) [inline docs/changelog.rst:5]:
```suggestion
* :feature:`1830` Bitcoin and Bitcoin Cash addresses are now derived from XPUBs keys when balances are refreshed.
```

**Review by yabirgb — COMMENTED** [review]
Some important comments

File: rotkehlchen/chain/manager.py (hunk 1 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/manager.py:625]: The order in which they're queried shouldn't change anything as the totals are eventually recalculated. I might be wrong though.

Yeah, you're right. It should only be done if btc/bch are queried.

File: rotkehlchen/chain/manager.py (hunk 2 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/manager.py:634]: Yes, you're right about that. Fixing atm

File: rotkehlchen/chain/manager.py [hunk 4]
```diff
@@ -621,10 +621,28 @@ def query_balances(
...
         self.totals = self.balances.recalculate_totals()
         return self.get_balances_update()
 
+    def derive_new_addresses_from_xpubs(
+            self,
+            should_derive_btc_xpubs: bool = False,
+            should_derive_bch_xpubs: bool = False,
+    ) -> None:
+        """Derive new addresses from xpubs.
+        """
```
Comment (yabirgb) [inline rotkehlchen/chain/manager.py:639]: Can be changed in a different PR
```suggestion
        """Derive new addresses from xpubs."""
```

**Review by yabirgb — COMMENTED** [review]
@LefterisJP take a look when you can

**Review by yabirgb — COMMENTED** [review]
Forgot to add this comment

File: rotkehlchen/chain/manager.py [hunk 5]
```diff
@@ -621,10 +621,28 @@ def query_balances(
             self.query_polkadot_balances(ignore_cache=ignore_cache)
         if should_query_avax:
             self.query_avax_balances(ignore_cache=ignore_cache)
+        if ignore_cache is True:
+            self.derive_new_addresses_from_xpubs(
+                should_derive_bch_xpubs=should_query_bch,
+                should_derive_btc_xpubs=should_query_btc,
+            )
 
         self.totals = self.balances.recalculate_totals()
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/manager.py:630]: @yabirgb what if no new addresses are derived? This would mean that there is a chance that we don't recalculate totals unless I am getting something incorrectly

[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/manager.py:630]: When does happen that the totals need to be recalculated if no new address is detected?

[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/manager.py:630]: Balances of all derived addresses are always updated (since we query them anyway when deriving addresses)

https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/bitcoin/xpub.py#L259

[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/manager.py:630]: Okey guess it is fine then

File: rotkehlchen/chain/manager.py (hunk 4 above)
Comment (LefterisJP) [reply rotkehlchen/chain/manager.py:639]: + 1 on this

**Review by LefterisJP — COMMENTED** [review]
Looks okayish but also needs `check_for_new_xpub_addresses` to have the appropriate logging for the respective bitcoin chain

**Review by LefterisJP — APPROVED** [review]
LGTM. Merging. So nice to see an all green
